service/src/main/java/bio/terra/cbas/dependencies/wes/CromwellClient.java

```diff
@@ -1,34 +1,69 @@
 package bio.terra.cbas.dependencies.wes;
 
+import bio.terra.cbas.common.exceptions.AzureAccessTokenException;
+import bio.terra.cbas.common.exceptions.DependencyNotAvailableException;
 import bio.terra.cbas.config.CromwellServerConfiguration;
+import bio.terra.cbas.dependencies.common.CredentialLoader;
+import bio.terra.cbas.dependencies.common.DependencyUrlLoader;
 import cromwell.client.ApiClient;
 import cromwell.client.api.EngineApi;
 import cromwell.client.api.Ga4GhWorkflowExecutionServiceWesAlphaPreviewApi;
 import java.util.Optional;
+import okhttp3.OkHttpClient;
 import org.springframework.stereotype.Component;
 
 @Component
 public class CromwellClient {
 
   private final CromwellServerConfiguration cromwellServerConfiguration;
+  private final DependencyUrlLoader dependencyUrlLoader;
+  private final CredentialLoader credentialLoader;
 
-  public CromwellClient(CromwellServerConfiguration cromwellServerConfiguration) {
+  private final OkHttpClient singletonHttpClient;
+
+  public CromwellClient(
+      CromwellServerConfiguration cromwellServerConfiguration,
+      DependencyUrlLoader dependencyUrlLoader,
+      CredentialLoader credentialLoader) {
     this.cromwellServerConfiguration = cromwellServerConfiguration;
+    this.dependencyUrlLoader = dependencyUrlLoader;
+    this.credentialLoader = credentialLoader;
+    singletonHttpClient = new ApiClient().getHttpClient();
   }
 
-  private ApiClient getApiClient() {
-    return new ApiClient().setBasePath(cromwellServerConfiguration.baseUri());
+  private ApiClient getApiClient()
+      throws DependencyNotAvailableException, AzureAccessTokenException {
+    String uri;
+
+    String baseUriFromConfig = cromwellServerConfiguration.baseUri();
+    if (!baseUriFromConfig.isEmpty()) {
+      uri = baseUriFromConfig;
+    } else {
+      uri = dependencyUrlLoader.loadDependencyUrl(DependencyUrlLoader.DependencyUrlType.CROMWELL);
+    }
+
+    ApiClient apiClient = new ApiClient().setBasePath(uri);
+    apiClient.setHttpClient(singletonHttpClient);
+    apiClient.addDefaultHeader(
+        "Authorization",
+        "Bearer " + credentialLoader.getCredential(CredentialLoader.CredentialType.AZURE_TOKEN));
+    // By closing the connection after each request, we avoid the problem of the open connection
+    // being force-closed ungracefully by the Azure Relay/Listener infrastructure:
+    apiClient.addDefaultHeader("Connection", "close");
```

Comment:
I think we only need this for the writer/external Cromwell instance... but I guess leaving it there for the local/reader instance is fine too

Reply:
It should be easy enough to move it into a writer function